### PR TITLE
Find consultation

### LINKED ISSUE #31: Consultation Booking
As a student, I want to create and join consultations, so that I can arrange group meetings with a lecturer without clashing with existing bookings.

Related Issues:
#32
#33
#35
#36
#38

### PR DESCRIPTION
**Pull request by BierVoetjie**
Closes #31

Summary: Add lecturer, date, and time search filters to the join consultation page with availability-aware empty states. When a filtered search returns no results, the page either prompts students to create a new consultation (when the date and time fall within the lecturer's availability) or informs them that the selection violates the lecturer's preferences. A compact inline Create button pre-fills the create consultation form with the selected lecturer and datetime.

## Key Changes

- **Search filters:** Added lecturer dropdown, date picker, and 15-minute-increment time picker to the join consultation page. Filters are applied server-side through `searchConsultationsForStudent`, which was extended with optional `lecturerId`, `date`, and `time` params while remaining backward-compatible with existing callers.
- **Empty states:** When filters return no results for a specific lecturer and date, the route calls `getLecturerAvailability` and `isDateAvailableForLecturer` to decide between two states: `'create'` (date and time are within the lecturer's availability window) or `'violation'` (outside the window).
- **New service function:** `isDateAvailableForLecturer` added to `consultation_availability_validation.js` — checks a date and optional start time against the lecturer's weekly slots, exception dates, and configured duration.
- **Pre-fill redirect:** The Create link passes `lecturerId`, `date`, and `time` as query params; `GET /consultations/new` already reads these to pre-fill the form's lecturer select and datetime-local input.
- **UI:** Empty-state Create button is styled as a compact inline element (`.button_compact`) rather than a full-width success button.
- **Time input constraint:** Client-side `updateTimeMin` prevents selecting a past time when today's date is chosen, and clears any previously entered time that would have become invalid.
- **Tests:** Unit tests for `getWeekdayFromIso` and `isDateAvailableForLecturer`; unit tests for `searchConsultationsForStudent` query construction; new `join_consultation` route unit test suite covering all GET filter paths and all POST join outcomes; integration tests for lecturer filter and both empty states; E2E test for the search → empty state → Create → pre-filled form flow.
- **Docs:** JSDoc added to all new and updated functions across `consultation_availability_validation.js`, `consultation_db.js`, `join_consultation.js`, and `join_consultation_search.js`.

## Acceptance

- Students searching by lecturer, date, or time see only matching future consultations.
- When no consultations match a specific lecturer and date that falls within the lecturer's availability, the page shows "No matching consultations. Create new consultation?" with a compact Create button inline.
- Clicking Create navigates to the create form with the lecturer and datetime pre-filled from the search.
- When the selected time falls outside the lecturer's availability window, the page shows "No matching consultation - Violates lecturer preferences."
- Selecting today's date prevents choosing a past time; any stale time value is automatically cleared.

## How to test locally

- Run full test suite: `npm test`
- Run only unit tests: `npm run test:unit`
- Run E2E (requires writable test MongoDB): `npm run test:e2e`

## Notes

- `buildLecturerOptions` is intentionally duplicated in `join_consultation.js` rather than extracted to a shared utility, to keep the route modules independently maintainable.
- `isDateAvailableForLecturer` is intentionally lighter than `validateLecturerAvailability` — it omits daily cap and overlap checks, which only apply at creation time.
- Integration and E2E tests for the empty states use a process-unique test lecturer ID to avoid interfering with seeded test data.

### CODE REVIEW

File: src/models/consultation_db.js [hunk 1]
```diff
@@ -54,12 +54,25 @@ const listConsultationsForLecturerOnDate = async function (lecturerId, isoDate)
...
 } = {}) {
-  const consultations = await consultationsCollection().find({}).toArray()
+  const query = {}
+  if (lecturerId) query.lecturerId = lecturerId
+  if (date && time) {
+    query.datetime = `${date}T${time}`
+  } else if (date) {
+    query.datetime = { $gte: `${date}T00:00`, $lt: `${date}T23:59~` }
+  }
+  const consultations = await consultationsCollection().find(query).toArray()
```
Comment (KheliD12345) [inline src/models/consultation_db.js:76]: using ~ for date matching can produce edge-case bugs around midnight. It is better to use an explicit end bound (or compute nextIsoDate and use $lt: nextIsoDate + 'T00:00') instead. Example:
const nextDate = new Date(isoDate);
nextDate.setDate(nextDate.getDate() + 1);
query.date = { $gte: `${isoDate}T00:00`, $lt: `${nextDate.toISOString().slice(0,10)}T00:00` };

**Review by KheliD12345 — APPROVED** [review]
Summary: Tests pass; good separation of concerns; added tests; clear JSDoc and input validation.

Issues: File: src/models/consultation_db.js — the date query uses a regex/operator like ~. Recommend replacing with an explicit end bound to avoid midnight/edge-case timestamps (better: compute the next day's midnight and use $lt).

Decision: Non-critical recommendations only. PR approved
